src/Agent.Worker/ResourceMetricsManager.cs

```diff
@@ -38,78 +47,329 @@ public void Setup(IExecutionContext context)
             }
             catch (Exception ex)
             {
-                return $"Unable to get CPU info, ex:{ex.Message}";
+                return StringUtil.Loc("ResourceMonitorCPUInfoError", ex.Message);
+            }
+        }
+
+        // Some compact Linux distributives like UBI may not have "free" utility installed,
+        // but we don't want to break currently existing pipelines, so ADO warning should be mitigated to the trace warning
+        public class MemoryMonitoringUtilityIsNotAvaliableException : Exception
+        {
+            public MemoryMonitoringUtilityIsNotAvaliableException(string message)
+                : base(message)
+            {
+            }
+        }
+
+        public struct MemoryInfo
+        {
+            public int TotalMemoryMB;
+            public int UsedMemoryMB;
+        }
+
+        public MemoryInfo GetMemoryInfo()
+        {
+            MemoryInfo memoryInfo = new();
+
+            ProcessStartInfo processStartInfo = new ProcessStartInfo();
+            var processStartInfoOutput = "";
+
+            if (PlatformUtil.RunningOnWindows)
+            {
+                processStartInfo.FileName = "wmic";
+                processStartInfo.Arguments = "OS GET FreePhysicalMemory,TotalVisibleMemorySize /Value";
+                processStartInfo.RedirectStandardOutput = true;
+
+                using (var process = Process.Start(processStartInfo))
+                {
+                    processStartInfoOutput = process.StandardOutput.ReadToEnd();
+                }
+
+                var processStartInfoOutputString = processStartInfoOutput.Trim().Split("\n");
+
+                var freeMemory = Int32.Parse(processStartInfoOutputString[0].Split("=", StringSplitOptions.RemoveEmptyEntries)[1]);
+                var totalMemory = Int32.Parse(processStartInfoOutputString[1].Split("=", StringSplitOptions.RemoveEmptyEntries)[1]);
+
+                memoryInfo.TotalMemoryMB = totalMemory / 1024;
+                memoryInfo.UsedMemoryMB = (totalMemory - freeMemory) / 1024;
             }
+
+            if (PlatformUtil.RunningOnLinux)
+            {
+                try
+                {
+                    processStartInfo.FileName = "free";
+                    processStartInfo.Arguments = "-m";
+                    processStartInfo.RedirectStandardOutput = true;
+
+                    using (var process = Process.Start(processStartInfo))
+                    {
+                        processStartInfoOutput = process.StandardOutput.ReadToEnd();
+                    }
+
+                    var processStartInfoOutputString = processStartInfoOutput.Split("\n");
+                    var memoryInfoString = processStartInfoOutputString[1].Split(" ", StringSplitOptions.RemoveEmptyEntries);
+
+                    if (memoryInfoString.Length != 7)
+                    {
+                        throw new MemoryMonitoringUtilityIsNotAvaliableException("Utility has non-default output");
```

Comment:
[not blocking/design feedback] If I follow correctly, the design here is that all the callers are expected to handle MemoryMonitoringUtilityIsNotAvaliableException.  That's not the best  design as if a caller does not handle MemoryMonitoringUtilityIsNotAvaliableException by accident, then an unexpected exception will be propagated

Instead to improve the design I'd suggest either:
1. pass a logger delegate into this function to allow it to write warnings directly OR
2. rename the function to TryGetMemoryInfo and use out parameters or an object result to return information to the caller to emit a warning

The pattern of defining a specific exception type only to catch it later it an anti-pattern of "using exceptions for control flow"

Reply:
Thanks for the comment! 
I agree with the idea to pass a logger delegate to this method, but it will require additional reworks in overall ResourceMetricsManager structure which seems unnecessary in the scope of the current monitor usage - to catch the resource usage spikes during the task execution. 
The general idea is that the monitor is not intended to be used outside of Task/Job Runners, where we have only two callers that safely handle the errors in `GetMemoryInfoString` method - it is the `RunMemoryUtilizationMonitor` and the `GetMemoryInfoString`.
But because the `GetMemoryInfo` method which produces the exception is not intended to be called from the outside of the monitor, I hid the internal methods to accord with the interface, and replaced the MemoryMonitoringUtilityIsNotAvaliableException with the general exception, the custom exception also was an overkill in a current usage scope.